### ISSUE TITLE
Type parameter variance inferred as invariant

### ISSUE THREAD

**Issue by daudef** [microsoft/pyright#6706]
In some cases Pyright infer `T` as invariant in then returns (rightfully) that the type `A[bool]` Cannot be assigned to `A[bool | None]`.

```python
import dataclasses


@dataclasses.dataclass
class A[T]:
    t: T


def works() -> A[bool | None]:
    return A(True)


def does_not_work() -> A[bool | None]:
    a = A(True)
    return a  # error on a

```

The error is 

```text
error: Expression of type "A[bool]" cannot be assigned to return type "A[bool | None]"
    "A[bool]" is incompatible with "A[bool | None]"
      Type parameter "T@A" is invariant, but "bool" is not the same as "bool | None" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```
I am using `pyright 1.1.339`.
I am not sure if it is an expected behavior or not, PEP 695 describe the inference algorithm, but I could not understand it in this case. 
But the fact that `works` works but `does_not_work` does not, seems weird to me.
Also I wonder if there is way to explicitly set a type parameter as covariant using the new syntax.

**Comment by erictraut** [comment]
Pyright's behavior here is correct. The type parameter `T` in class `A` is invariant because it's used for a mutable (readable and writable) value. If you change the class to a "frozen" dataclass, then `t` is no longer mutable, and pyright will correctly infer that `T` is covariant.

```python
@dataclasses.dataclass(frozen=True)
class A[T]: # T is covariant in this case
    t: T
```

There's no way to explicitly set a type variable as covariant with the new syntax. You could use the old syntax to do so, but it would be incorrect (not type safe) to use a covariant type parameter in class `A` if the dataclass is not frozen.

The type argument for a generic class is established at the time you construct an instance of that class. You can specify the type argument explicitly (as in `A[bool | None](True)`). If you don't specialize the class explicitly, pyright will use type inference when evaluating the constructor call to determine the type argument. If there is additional context, such as the "expected type" provided by the `return` statement in `works`, pyright uses bidirectional type inference. If no additional context is provided, as in the `does_not_work` example, then normal type inference rules apply.

For more information about bidirectional type inference, refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types).

**Comment by daudef** [comment]
Thanks a lot Eric, your explanation is super clear.
